src/test/txrequest_tests.cpp

```diff
@@ -0,0 +1,237 @@
+        tracker.RequestedTx(peers[0], txid, now);
+        now += TEST_TIMEOUT;
+    } else if (remove_reason == 2) {
+        tracker.RequestedTx(peers[0], txid, now);
+        tracker.ReceivedResponse(peers[0], txid);
+    }
+
+    CheckRequestable(tracker, peers[0], now, {}); // If peer 0 goes offline, or the request times out, or a NOTFOUND is received, does peer 1 gets it (despite peer 2 being earlier)
+    CheckRequestable(tracker, peers[1], now, {txid});
+    CheckRequestable(tracker, peers[2], now, {});
+    if (remove_reason == 0) {
+        tracker.DeletedPeer(peers[1]);
+    } else if (remove_reason == 1) {
+        tracker.RequestedTx(peers[1], txid, now);
+        now += TEST_TIMEOUT;
+    } else if (remove_reason == 2) {
+        tracker.RequestedTx(peers[1], txid, now);
+        tracker.ReceivedResponse(peers[1], txid);
+    }
+    CheckRequestable(tracker, peers[0], now, {}); // Only if peer 1 also goes offline, or the request times out, does peer 2 get it.
+    CheckRequestable(tracker, peers[1], now, {});
+    CheckRequestable(tracker, peers[2], now, {txid});
+}
+
+void TestOrderDependsOnPeers(bool delay)
+{
+    TxRequestTracker tracker(TEST_TIMEOUT);
+    auto peers = GenPeers(3);
+    auto txid1 = GenTxid(tracker, {peers[1], peers[2]}); // A txid such that peer 1 is prioritized over peer 2
+    auto txid2 = GenTxid(tracker, {peers[2], peers[1]}); // A txid such that peer 2 is prioritized over peer 1
+    auto now = RandomTime();
+
+    for (auto peer : peers) {
+        tracker.ReceivedInv(peer, txid1, ReqTime(now, delay)); // Announce both txids for all peers
+        tracker.ReceivedInv(peer, txid2, ReqTime(now, delay));
+    }
+    if (delay) now += TEST_DELAY;
+    CheckRequestable(tracker, peers[0], now, {txid1, txid2}); // Both are requestable
+    tracker.DeletedPeer(peers[0]); // Peer 0 goes offline
+    CheckRequestable(tracker, peers[1], now, {txid1}); // txid1 is now requestable from peer 1
+    CheckRequestable(tracker, peers[2], now, {txid2}); // While txid2 is now requestable from peer 2
+}
+
+void TestRequestableOrder(bool delay)
+{
+    TxRequestTracker tracker(TEST_TIMEOUT);
+    auto peers = GenPeers(2);
+    auto txid1 = GenTxid(tracker, {});
+    auto txid2 = GenTxid(tracker, {});
+    auto now = RandomTime();
+
+    tracker.ReceivedInv(peers[0], txid1, ReqTime(now, delay));
+    tracker.ReceivedInv(peers[0], txid2, ReqTime(now, delay));
+    tracker.ReceivedInv(peers[1], txid2, ReqTime(now, delay));
+    now -= std::chrono::microseconds{1};
+    tracker.ReceivedInv(peers[1], txid1, ReqTime(now, delay));
+    if (delay) now += TEST_DELAY;
+    now += std::chrono::microseconds{1};
+    CheckRequestable(tracker, peers[0], now, {txid1, txid2}); // Check that for each peer, requests go out in announcement order,
+    tracker.DeletedPeer(peers[0]);
+    CheckRequestable(tracker, peers[1], now, {txid2, txid1}); // even when the clock went backwards during announcements.
+}
+
+void TestAll()
```

Comment:
75fd5dc

I think `AlreadyHaveTx` and time-goes-backward of `SetTimePoint` aren't covered? I'll try to add them soon, also `ReceivedResponse` should be called unconditionally at the end of each test to verify clean state ?

Reply:
The unit tests are completely rewritten. There aren't any for the time-moving-backward though, I'll try to add that soon.

Reply:
Actually, that's not necessary anymore. As `ReceivedInv` doesn't have a "now" argument anymore, the time of announcement clearly doesn't matter, simply by the API being agnostic of it. I'm adding a test that has a reqtime that decreases, though.